## Part 1: The Doomsday Algorithm

The Doomsday algorithm, devised by mathematician J. H. Conway, computes the day of the week any given date fell on. The algorithm is designed to be simple enough to memorize and use for mental calculation.

__Example.__ With the algorithm, we can compute that July 4, 1776 (the day the United States declared independence from Great Britain) was a Thursday.

The algorithm is based on the fact that for any year, several dates always fall on the same day of the week, called the <em style="color:#F00">doomsday</em> for the year. These dates include 4/4, 6/6, 8/8, 10/10, and 12/12.

__Example.__ The doomsday for 2016 is Monday, so in 2016 the dates above all fell on Mondays. The doomsday for 2017 is Tuesday, so in 2017 the dates above will all fall on Tuesdays.

The doomsday algorithm has three major steps:

1. Compute the anchor day for the target century.
2. Compute the doomsday for the target year based on the anchor day.
3. Determine the day of week for the target date by counting the number of days to the nearest doomsday.

Each step is explained in detail below.

### The Anchor Day

The doomsday for the first year in a century is called the <em style="color:#F00">anchor day</em> for that century. The anchor day is needed to compute the doomsday for any other year in that century. The anchor day for a century $c$ can be computed with the formula:
$$
a = \bigl( 5 (c \bmod 4) + 2 \bigr) \bmod 7
$$
The result $a$ corresponds to a day of the week, starting with $0$ for Sunday and ending with $6$ for Saturday.

__Note.__ The modulo operation $(x \bmod y)$ finds the remainder after dividing $x$ by $y$. For instance, $12 \bmod 3 = 0$ since the remainder after dividing $12$ by $3$ is $0$. Similarly, $11 \bmod 7 = 4$, since the remainder after dividing $11$ by $7$ is $4$.

__Example.__ Suppose the target year is 1954, so the century is $c = 19$. Plugging this into the formula gives
$$a = \bigl( 5 (19 \bmod 4) + 2 \bigr) \bmod 7 = \bigl( 5(3) + 2 \bigr) \bmod 7 = 3.$$
In other words, the anchor day for 1900-1999 is Wednesday, which is also the doomsday for 1900.

__Exercise 1.1.__ Write a function that accepts a year as input and computes the anchor day for that year's century. The modulo operator `%` and functions in the `math` module may be useful. Document your function with a docstring and test your function for a few different years.  Do this in a new cell below this one.

In [2]:
def get_anchorday(year):
    '''This functions is used to calculate the anchor day of a century where the given year belongs to.

    Input:
        year: an integer, ie(320,1864,etc.)

    Output:
        a tuple contains the value of a and anchor day, ie(0, Sunday)
    '''
    year_string = str(year)
    c = int(year_string[:-2])    
    a =(5 * (c % 4) + 2) % 7
    day_of_week = {0:'Sunday', 1:'Monday', 2:'Tuesday', 3:'Wednesday', 4:'Thursday', 5:'Friday', 6:'Saturday'}
    return (a, day_of_week[a])

print get_anchorday(320)
print get_anchorday(1987)
print get_anchorday(1830)

(3, 'Wednesday')
(3, 'Wednesday')
(5, 'Friday')


### The Doomsday

Once the anchor day is known, let $y$ be the last two digits of the target year. Then the doomsday for the target year can be computed with the formula:
$$d = \left(y + \left\lfloor\frac{y}{4}\right\rfloor + a\right) \bmod 7$$
The result $d$ corresponds to a day of the week.

__Note.__ The floor operation $\lfloor x \rfloor$ rounds $x$ down to the nearest integer. For instance, $\lfloor 3.1 \rfloor = 3$ and $\lfloor 3.8 \rfloor = 3$.

__Example.__ Again suppose the target year is 1954. Then the anchor day is $a = 3$, and $y = 54$, so the formula gives
$$
d = \left(54 + \left\lfloor\frac{54}{4}\right\rfloor + 3\right) \bmod 7 = (54 + 13 + 3) \bmod 7 = 0.
$$
Thus the doomsday for 1954 is Sunday.

__Exercise 1.2.__ Write a function that accepts a year as input and computes the doomsday for that year. Your function may need to call the function you wrote in exercise 1.1. Make sure to document and test your function.

In [3]:
def get_doomsday(year):
    '''This functions is used to calculate the doomsday of a given year.

    Input:
        year: an integer, ie(320,1864,etc.)

    Output:
        a tuple contains the value of 'd' and doomsday, ie(0, Sunday)
    '''    
    a = get_anchorday(320)[0]    # calculate the anchor day
    year_string = str(year)
    y = int(year_string[-2:])    # last two digits
    d = int((y + round(y/4) + a) % 7)
    day_of_week = {0:'Sunday', 1:'Monday', 2:'Tuesday', 3:'Wednesday', 4:'Thursday', 5:'Friday', 6:'Saturday'}
    return (d, day_of_week[d])

print get_doomsday(1954)
print get_doomsday(1824)

(0, 'Sunday')
(5, 'Friday')


### The Day of Week

The final step in the Doomsday algorithm is to count the number of days between the target date and a nearby doomsday, modulo 7. This gives the day of the week.

Every month has at least one doomsday:
* (regular years) 1/10, 2/28
* (leap years) 1/11, 2/29
* 3/21, 4/4, 5/9, 6/6, 7/11, 8/8, 9/5, 10/10, 11/7, 12/12

__Example.__ Suppose we want to find the day of the week for 7/21/1954. The doomsday for 1954 is Sunday, and a nearby doomsday is 7/11. There are 10 days in July between 7/11 and 7/21. Since $10 \bmod 7 = 3$, the date 7/21/1954 falls 3 days after a Sunday, on a Wednesday.

__Exercise 1.3.__ Write a function to determine the day of the week for a given day, month, and year. Be careful of leap years! Your function should return a string such as "Thursday" rather than a number. As usual, document and test your code.

In [4]:
def get_day_of_week(year, month, day):
    '''This functions is used to calculate the day of a week for a given day.

    Input:
        year: an integer, ie(320,1864,etc.)
        month: an integer, ie(1,2,3...)
        day: an integer, ie(12,24...)
    Output:
        today: a string of the day of week for today. ie('Sunday')
    '''  
    # determine whether the given year is a leap year
    if (str(year)[-2:] == '00' and year % 400 == 0) or (str(year)[-2:] != '00' and year % 4 == 0):
        leap = 1
    else:
        leap = 0
    # two tuples of the doomsdays.
    doom_list_leap = (11, 29, 21, 4, 9, 6, 11, 8, 5, 10, 7, 12)
    doom_list_noleap = (10, 28, 21, 4, 9, 6, 11, 8, 5, 10, 7, 12)
    doomsday = get_doomsday(year)[0]
    # Find the the target day which is the doomsday of that month
    if leap == 1:
        target_day = doom_list_leap[month-1]
    else:
        target_day = doom_list_noleap[month-1]
    
    diff = day - target_day
    
    if diff >= 0:
        remain_in_week = diff % 7 # if diff >= 0, remain_in_week >0. Otherwise, remain_in week < 0
    else:
        remain_in_week = -(abs(diff) % 7)
    
    today_index = (doomsday + remain_in_week) % 7 # index of today
    day_of_week = {0:'Sunday', 1:'Monday', 2:'Tuesday', 3:'Wednesday', 4:'Thursday', 5:'Friday', 6:'Saturday'}
    today = day_of_week[today_index]
    
    return today

print get_day_of_week(1954, 7, 10)
print get_day_of_week(1954, 7, 11)
print get_day_of_week(1954, 7, 21)

Saturday
Sunday
Wednesday


__Exercise 1.4.__ How many times did Friday the 13th occur in the years 1900-1999? Does this number seem to be similar to other centuries?

In [10]:
def get_13_century(century_start):
    ''' Calculate the times that Friday the 13th occur in a century
        
    Arguments:
        century_start(int): The beginning year of a century. ie(1900 for 1900 - 1999; 1800 for 18th century)
    
    Returns:
        print the count on the screen
    '''
    count = 0
    
    for year in range(century_start,century_start + 100):
        for month in range(1,13):
            if get_day_of_week(year, month, 13) == 'Friday':
                count += 1

    print count

get_13_century(1900)

172


<h3>There are 172 Friday the 13th occur in the years 1900-1999</h3>

In [11]:
get_13_century(1900)
get_13_century(1800)
get_13_century(1600)
get_13_century(1200)

172
172
173
173


<h3>Friday the 13th occurs 173 times in all leap years and 172 time in all regular years</h3>

__Exercise 1.5.__ How many times did Friday the 13th occur between the year 2000 and today?

In [27]:
count = 0
# all 13th occuring between 2000 and 2016
for year in range(2000,2017):
    for month in range(1,13):
        if get_day_of_week(year, month, 13) == 'Friday':
            count += 1
# By now, only 01/13/2017 occurs in 2017.
if get_day_of_week(2017, 1, 13) == 'Friday':
            count += 1
print count

30


## Part 2: 1978 Birthdays

__Exercise 2.1.__ The file `birthdays.txt` contains the number of births in the United States for each day in 1978. Inspect the file to determine the format. Note that columns are separated by the tab character, which can be entered in Python as `\t`. Write a function that uses iterators and list comprehensions with the string methods `split()` and `strip()` to  convert each line of data to the list format

```Python
[month, day, year, count]
```
The elements of this list should be integers, not strings. The function `read_birthdays` provided below will help you load the file.

In [97]:
def read_birthdays(file_path):
    """Read the contents of the birthdays file into a string.
    
    Arguments:
        file_path (string): The path to the birthdays file.
        
    Returns:
        string: The contents of the birthdays file.
    """
    with open(file_path) as file:
        return file.read()


def birthday_split(file_path):
    '''This is used to load the txt file and split the birthday into single element
    
    Arguments:
        file_path (string): The path to the birthdays file.
    
    Returns:
        birthday_final_integer(list): A list of lists in which every small list looks like [month, day, year, count]
    '''
    birthday = read_birthdays(file_path)
    birthday = birthday.strip()            # get rid of the spaces in front of and behind the main text
    birthday = birthday.split('\n')[6:]    # split the birthdays by '\n' and remove the title lines
    birthday =[i.split('/') for i in birthday]  # First split by '/'
    
    birthday_clean = []
    
    for i in birthday:
        birthday_clean.append( [j.split('\t') for j in i] )    # Second split by '\t'
    
    birthday_final = map(lambda x: sum(x, []), birthday_clean) # merge the small lists within a big list
    
    for i in birthday_final:    # change the two digits year into four digits style.
        i[2] = '19' + i[2]
    
    birthday_final_integer = [map(int, i) for i in birthday_final] # change the strings into integer.
    
    return birthday_final_integer       

# For test        
file = './birthdays.txt'
birthday = birthday_split(file)
print len(birthday)
birthday[:6]

365


[[1, 1, 1978, 7701],
 [1, 2, 1978, 7527],
 [1, 3, 1978, 8825],
 [1, 4, 1978, 8859],
 [1, 5, 1978, 9043],
 [1, 6, 1978, 9208]]

__Exercise 2.2.__ Which month had the most births in 1978? Which day of the week had the most births? Which day of the week had the fewest? What conclusions can you draw? You may find the `Counter` class in the `collections` module useful.

In [101]:
def month_most_year(birthdays):
    '''Calculate the month had the most births in 1978 
     
     Arguments: 
         birthdays(list): We got it from function birthday_split
     
     Returns:
         integer: the month had the most births in 1978
     '''
    month_count = [0] * 12
    
    for i in birthdays:
        month_count[i[0] - 1] += i[3]
    
    return month_count.index(max(month_count)) + 1

print month_most_year(birthday)

8


In [123]:
def day_of_week_most_year(birthdays, big_or_small):
    '''Calculate the day of a week had the most births in 1978 
     
     Arguments: 
         birthdays(list): We got it from function birthday_split
         big_or_small(string): indicate the output should by largest or smallest. ie(most or fewest)
     Returns:
         string: the day of a week had the most births in 1978
     '''
    day_count = {'Sunday':0, 'Monday':0 , 'Tuesday':0, 'Wednesday':0, 'Thursday':0, 'Friday':0, 'Saturday':0}
    
    for i in birthdays:
        day = get_day_of_week(i[2], i[0], i[1])
        day_count[day] += i[3]
    
    if big_or_small == 'most':
        return max(day_count, key = lambda i: day_count[i])
    else:
        return min(day_count, key = lambda i: day_count[i])

print day_of_week_most_year(birthday, 'most')
print day_of_week_most_year(birthday, 'fewest')

Tuesday
Sunday


<h3>August had most births in 1978, Tuesday had the most births while Sunday had the fewest in 1978</h3>
<h3>Births always occured in summer and weekdays instead of weekends</h3>

__Exercise 2.3.__ What would be an effective way to present the information in exercise 2.2? You don't need to write any code for this exercise, just discuss what you would do.

<h3>I think pie chart and hisogram are the best ways to present the distribution of births in 1978 over month and day of week</h3>